### PR TITLE
datapath: probe socket match support, plumb to Envoy configuration

### PR DESCRIPTION
**Pull request by jrajahalme**
Let Envoy know if it may use original source addresses in upstream
connections. Limit support to egress proxy only for now.

This fixes policy regressions when the iptables socket match kernel
module is not available by not using original source address in proxy
upstream connections in that case.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9016)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

File: pkg/datapath/iptables/iptables.go [hunk 1]
```diff
@@ -334,7 +334,16 @@ func (m *IptablesManager) Init() {
 			"ip6tables kernel modules could not be loaded, so IPv6 cannot be used")
 		ip6tables = false
 	}
-	m.ip6tables = ip6tables
+	m.haveIp6tables = ip6tables
+
+	if err := modulesManager.FindOrLoadModules("xt_socket"); err != nil {
```
Comment (brb) [inline pkg/datapath/iptables/iptables.go:339]: Shouldn't we probe for the `nf_socket_ipv{4,6}` modules as well?

```
> lsmod | grep xt_socket
xt_socket              16384  0
nf_socket_ipv4         16384  1 xt_socket
nf_socket_ipv6         20480  1 xt_socket
nf_defrag_ipv6         24576  2 nf_conntrack,xt_socket
nf_defrag_ipv4         16384  2 nf_conntrack,xt_socket
x_tables               49152  8 xt_conntrack,iptable_filter,xt_socket,xt_tcpudp,xt_addrtype,xt_nat,ip_tables,xt_MASQUERADE
```

Comment (joestringer) [reply pkg/datapath/iptables/iptables.go:339]: Given that they are dependencies, I think that if they are not available then `xt_socket` load will fail anyway. Unless you're concerned about kernels with IPv6 compiled out?

Comment (brb) [reply pkg/datapath/iptables/iptables.go:339]: Makes sense.

**Comment by jrajahalme** [comment]
Unrelated test flake when contacting an outside resource (www.google.com), retesting.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Testing the effect of missing socket match support in #9001, prior to this PR it resulted in two egress to world test failures.

**Comment by jrajahalme** [comment]
Rebased to resolve merge conflicts.

**Comment by jrajahalme** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**